aten/src/ATen/native/ScatterGather.cpp

```diff
@@ -0,0 +1,82 @@
+#include <tuple>
+#include <algorithm>
+#include <ATen/ATen.h>
+#include <c10/core/WrapDimMinimal.h>
+
+namespace at { namespace native {
+
+Tensor & gather_out(Tensor & result, const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  return at::_gather_out(result, self, dim, index, sparse_grad);
+}
+
+Tensor gather(const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  return at::_gather(self, dim, index, sparse_grad);
+}
+
+Tensor & scatter_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  return at::_scatter_(self, dim, index, source);
+}
+
+Tensor & scatter_(Tensor & self, int64_t dim, const Tensor & index, Scalar value) {
+  return at::_scatter_(self, dim, index, value);
+}
+
+Tensor scatter(const Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  Tensor ret = self.clone();
+  return at::_scatter_(ret, dim, index, source);
+}
+
+Tensor scatter(const Tensor & self, int64_t dim, const Tensor & index, Scalar value) {
+  Tensor ret = self.clone();
+  return at::_scatter_(ret, dim, index, value);
+}
+
+Tensor & scatter_add_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  if (index.numel() == 0) {
```

Comment:
please comment this function -- this code didn't exist before and I'm not 100% sure why you need it?  (For this if for example, I think it's wrong because of the NoScalars checks in the existing code and you are just matching the behavior, but I'm just guessing.)

Reply:
I understand this looks awful, but the behavior of scatter on empty index is another undefined and uncheck area. Please see this test:
![image](https://user-images.githubusercontent.com/1032377/62507069-0a473f00-b7d1-11e9-9d9d-bf6ccce6537d.png)
The document says 
> self, index and other should have same number of dimensions. 

But they are not checked. Clearly, we can see that some of the above illegal `index` get accepted by mistake. Since the purpose of this PR is not to fix this issue, I will not do much on this here. But I will modify the implementation so that they behave exactly as the original one, and add a quick comment saying they are the special cases handled in the original one.

Reply:
ya, that's what I expected, thanks for checking.  The issue is that the original function doesn't actually look at the sizes, it looks at the "legacy sizes."

Reply:
hmm, I think the complexity cost here is pretty high (basically, reproducing the incorrect TH behavior in ATen).  Can you either:
1) (preferably) port scatter/scatterAdd to ATen
2) (not as good) just write a scatter_add scalar version in TH, e.g. update Declarations.cwrap

Reply:
@gchanan Yes, I think it makes sense to port things from TH to ATen. The implementation of broadcasting (calling `_gather` from `gather`) in https://github.com/pytorch/pytorch/pull/23479 is also ugly.

Reply:
@zasdfgbnm can you help me understand your current plan?  Is it something like:
1) port everything to ATen
2) add scalar support for scatter_add_
3) add broadcasting support

or something else?

Reply:
@gchanan Similar to what you said, but two stacks of PRs:

PRs for gather
1. Port gather to ATen
2. Broadcasting support

PRs for scatter:
1. Port scatter and scatter_add to ATen
2. Fix docs and error checking, basically the No. 3 in https://github.com/pytorch/pytorch/pull/23510
3. Add scalar support for scatter_add_
4. Broadcasting support